### ISSUE TITLE
Reproducing parallel results

### ISSUE THREAD

**Issue by mortont** [bfs18/nsynth_wavenet#8]
I've been trying to reproduce the parallel wavenet results, however I'm running into some issues with training the teacher model. I have trained it on the LJ Speech dataset with the default wavenet_mol.json configuration to ~280k steps (all other hparams are unchanged as well). The loss looks good, however the evaluated speech is just babbling, as if local conditioning wasn't used.

I didn't see anything immediately apparent as to why this is happening, do you have any ideas?

**Comment by bfs18** [comment]
Hi @mortont 
Could you provide more information? For example, the loss curve, some generated samples and the code commit.

**Comment by mortont** [comment]
Sure, I'm on commit `67eacb995aef465d1e2ed810f25e0d7d3899e9b6` and this is the loss curve 
<img width="922" alt="screen shot 2018-07-13 at 9 33 06 am" src="https://user-images.githubusercontent.com/12803317/42694285-de28624c-867f-11e8-8711-95a56b332f9c.png">

This is one of the generated samples: [gen_LJ001-0028.wav.zip](https://github.com/bfs18/nsynth_wavenet/files/2192844/gen_LJ001-0028.wav.zip) 
For reference, this is the original that was used for the mel conditioning: [LJ001-0028.wav.zip](https://github.com/bfs18/nsynth_wavenet/files/2192846/LJ001-0028.wav.zip)

What loss did you let the teacher get to in your examples? Wondering if this just needs more training time. Let me know if any other information would be helpful.

**Comment by bfs18** [comment]
According to my experience, if a small batch size such as 2 is used, 200k training steps is not enough. Because the model haven't seen enough data. What's your batch size? Besides please ensure USE_RESIZE_CONV=False in masked.py.
This is my loss curve.
![loss_curve](https://user-images.githubusercontent.com/39617376/42696001-8cb390a2-86e9-11e8-9c2e-036faf795fed.png)

**Comment by mortont** [comment]
Ah, thank you. My batch size is 2 so I'll continue training. I'll go ahead and close this issue and report back when I have good samples for posterity.

**Comment by bfs18** [comment]
Hi @mortont , I'm sorry that the wavenet results cannot be reproduced with the default config json. Recently I cannot reporduce the previous result either. And I finally figure out that weight normalization harms the wavenet performance. So setting use_weigh_norm=false in wavenet_mol.json may solve your problem. 
Initially weight normalization produces some promising results for  parallel wavenet. So I keep it as a default configuration. Probabily it has negative effects on the learned mel-sepctrum representation. And good mel-spectrum representation is vital for a good modle, since the wave sample points totaly depend on theit. 
I'm also doing simaliar tests on parallel wavenet.
Once again, I'm sorry.

**Comment by bfs18** [comment]
Hi, 
I tested the code, setting `use_weigth_norm=False` solves the problem.
And I implemented http://export.arxiv.org/abs/1807.07281 at this weekend. I will update the code after some testing.

**Comment by bfs18** [comment]
Hi @zhang-jian, I'm sorry for that.
The first line of ns_pwn-eval/checkpoint contains the absolute path of the model checkpoint. You can modify the path according to your file system. Or you can only keep the basename of that path. Refer to ns_wn-eval/checkpoint for an example.

**Comment by mortont** [comment]
I was able to confirm that the pretrained model produces a recognizable voice, all that was needed was changing the path in the checkpoint file to a relative one. Great work!

@bfs18 ClariNet looks very interesting, mostly because of the more numerically stable implementation. I've noticed that parallel wavenet optimization is very difficult and unstable, so hopefully ClariNet helps with that.

**Comment by bfs18** [comment]
I got some new examples running with contrastive loss and without weight normalization at step 70948. The result may improve a bit after longer running.
[gen_LJ001-0001](https://drive.google.com/open?id=1Egw9l8iT89DamojIhGJEr98d4FxSDTq7)
[gen_LJ001-0001](https://drive.google.com/open?id=1TYGCsWaS6kxV5ijg-YfEpTGeegsyFU0c)

![loss4](https://user-images.githubusercontent.com/39617376/43375093-91fc18d2-93e5-11e8-9e5b-ad040f696a80.png)
![loss3](https://user-images.githubusercontent.com/39617376/43375094-92268f2c-93e5-11e8-8806-1afd6e4b3df9.png)
![loss2](https://user-images.githubusercontent.com/39617376/43375095-92506626-93e5-11e8-88c5-ea2763035703.png)
![loss1](https://user-images.githubusercontent.com/39617376/43375096-92793ad8-93e5-11e8-8208-c16ffcc09020.png)

**Comment by bfs18** [comment]
I look deeper into the problem.
Some bad configuration (e.g. weight normalization + tanh trans conv act) may cause the activation of the transposed convolution layer saturate. So the mel condition becomes meaningless. The model degenerate to an unconditional one.
The following figures are the histogram and spectrum of the transposed convolution stack output. This model only generate random speech even though mel condition is used.
![after_act2_hist](https://user-images.githubusercontent.com/39617376/43434728-89532df4-94af-11e8-8b31-8f15b748f34e.png)
![after_act2](https://user-images.githubusercontent.com/39617376/43434730-897e2568-94af-11e8-8173-36bcec4145fe.png)
In contrast, the following figures come from an OK model.
![after_act2_hist](https://user-images.githubusercontent.com/39617376/43434800-dab90380-94af-11e8-8434-cadae31f1f1c.png)
![after_act2](https://user-images.githubusercontent.com/39617376/43434801-dae6d3f0-94af-11e8-92b1-e5910626dd06.png)
Most of the activation values are close to 0. So the learned representation may be considered sparse.
I think there are 2 solutions. 
1. use activation functions that would not saturate, e.g. leaky_relu in Clarinet.
2. use dropout to make the teacher forcing input incomplete. So the model is forced to access the additional mel condition data to predict the original waveform. I'm not sure to which layers dropout should be applied. I am working on this.

**Comment by EdisonChen726** [comment]
Hi,
I am running the eval_parallel_wavenet.py, after 60K training, it can generate the audio with content, however, the sound is quite light, is this problem related to the power loss. Beside, the config does not include contrastive loss, how should I set this parameter?

**Comment by bfs18** [comment]
Hi @EdisonChen726 
I uploaded the model with contrastive loss. You can find the configuration json in the package.
https://drive.google.com/open?id=1AtofQdXbSutb-_ZWFeA_I17NR2i8nUC7

**Comment by EdisonChen726** [comment]
@bfs18 thank you for the fast reply, I will try it asap

**Comment by bfs18** [comment]
Updated Clarinet vocoder results. Clarinet results has similar noise compared to pwn results. So I think the noise comes from the power loss term.
Comapred to the teacher result, the student result does not have clear formats between 1000Hz to 3000Hz. This may be the source of the noise in the waves generated by the student.
teacher spec
![wn1](https://user-images.githubusercontent.com/39617376/43772917-706469ee-9a76-11e8-863a-5a4bc9a7583b.jpg)
student spec
![pwn1](https://user-images.githubusercontent.com/39617376/43772916-7022ccc8-9a76-11e8-9c4a-840f3301d2d5.jpg)
The priority frequencies loss implemented in [keithito/tacotron](https://github.com/keithito/tacotron/blob/daae2d915bec320be5b0785a1c0121a9b6e85ad4/models/tacotron.py#L119) may relivate the problem.

**Comment by EdisonChen726** [comment]
@bfs18 hi, have you meet the problem of the very slight audio result, I need to turn up the volume to very high so that I can hear the voice, do you have any idea why this happen?

**Comment by bfs18** [comment]
Hi @EdisonChen726 
Setting use_mu_law=True would cause low volume when training parallel wavenet.  
It is caused by clip_quant_scale function in L13 in wavenet/parallelgen.py. I don't know how to solve the problem.
You can test it with the following code.
```
import numpy as np
import librosa 

def inv_mu_law_numpy(x, mu=255.0):
    x = np.array(x).astype(np.float32)
    out = (x + 0.5) * 2. / (mu + 1)
    out = np.sign(out) / mu * ((1 + mu) ** np.abs(out) - 1)
    out = np.where(np.equal(x, 0), x, out)
    return out

def cast_quantize_numpy(x, quant_chann):
    x_quantized = x * quant_chann / 2
    return x_quantized.astype(np.int32)

audio, _ = librosa.load('test_data/test.wav', sr=16000)
audio_int = cast_quantize_numpy(audio, 2 ** 8)
audio_ = inv_mu_law_numpy(audio_int)
librosa.output.write_wav('test_data/test_inv.wav', audio_, sr=16000)
```
The volume of the output wave becomes very low.

**Comment by EdisonChen726** [comment]
@bfs18 get it! thank you so much. 
Right now I have another problem, when I add the  contrastive loss with 0.3, there will be the error as:
ResourceExhaustedError (see above for traceback): OOM when allocating tensor with shape[4,64,1,7680] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
Do you know how to solve it?

**Comment by bfs18** [comment]
@EdisonChen726  just try smaller batch size.

**Comment by switchzts** [comment]
@bfs18 Hi,I use your model to test synthesized speech, the mute part will become a murmur, and the non-mute part is normal. Do you know why? Is it because of the trim at the time of training?

**Comment by bfs18** [comment]
Hi @switchzts 
The use_mu_law + mol waves are much cleaner at mute part. However no_mu_law + mol waves are just as you say. So 200k steps may be not enough to train a good no_mu_law + mol model. I am not sure whether trim is a problem.

**Comment by EdisonChen726** [comment]
@bfs18 Hi, I tried to set the batch size as 1, but the same error happened.

**Comment by HallidayReadyOne** [comment]
Hi @bfs18 , I have some questions about initialization. 
1. Why the scale_params bias init by -0.3?  Is this an experience value?  And why not use log scale in student net?
https://github.com/bfs18/nsynth_wavenet/blob/4370294c8c088d3bc9e8b8486e75af9fe7f845cb/wavenet/parallel_wavenet.py#L243
https://github.com/bfs18/nsynth_wavenet/blob/4370294c8c088d3bc9e8b8486e75af9fe7f845cb/wavenet/parallel_wavenet.py#L92

2. In readme, you mentioned
"Proper initial **mean_tot** and **scale_tot** values have positive impact on model convergence and numerical stability. According to the LJSpeech data distribution, proper initial values for mean_tot and scale_tot should be 0.0 and 0.05. **I modified the initializer to achieve it**."
In parallel_wavenet.py **mean_tot** and **scale_tot** are init by 0 and 1, which initializer is modified to achieve proper initial values for mean_tot and scale_tot (0.0 and 0.05).
https://github.com/bfs18/nsynth_wavenet/blob/4370294c8c088d3bc9e8b8486e75af9fe7f845cb/wavenet/parallel_wavenet.py#L276

Thank you!

**Comment by bfs18** [comment]
Hi @EdisonChen726 
What's you gpu memory size? I only run the code on gpu with 12 gb memory or more.

**Comment by bfs18** [comment]
Hi @HallidayReadyOne 
--Why the scale_params bias init by -0.3? Is this an experience value?
Yes.
I wrote some memos on why chose this value and why not use log scale in the comments in [test_scale](https://github.com/bfs18/nsynth_wavenet/blob/master/tests/test_scale.py)
Let me know if you need further  explaination.

**Comment by bfs18** [comment]
Hi @HallidayReadyOne 
You are right this value is picked when use_weight_norm=True. Since it is chosen by experience, it is not that strict. When setting use_weight_norm=False, the initial scale is still small enough. So I keep this value.

**Comment by HallidayReadyOne** [comment]
Thanks @bfs18, another question about init is,
In readme, you mentioned
"Proper initial mean_tot and scale_tot values have positive impact on model convergence and numerical stability. According to the LJSpeech data distribution, proper initial values for mean_tot and scale_tot should be 0.0 and 0.05. I modified the initializer to achieve it."
Could you please explain a little bit about how this is achieved?

**Comment by EdisonChen726** [comment]
@bfs18 my gpu memory size is also 12GB

**Comment by HallidayReadyOne** [comment]
Hi @bfs18 , you mentioned earlier that **setting use_mu_law=True would cause low volume when training parallel wavenet** (I also encountered the same problem).

This may not be caused by the clip_quant_scale function in L13 in wavenet/parallelgen.py?

I test it with the following code:

audio, _ = librosa.load('test_data/test.wav', sr=16000)
**audio_mu = mu_law_numpy(audio)
audio_mu_scaled = np.asarray(audio_mu, np.float32) / (2 ** 8 / 2.)**
audio_int = cast_quantize_numpy(audio_mu_scaled, 2 ** 8)
audio_ = inv_mu_law_numpy(audio_int)
librosa.output.write_wav('test_data/test_inv.wav', audio_, sr=16000)

Because when use_mu_law=True set to true, the output of the student model is scaled mu-law data, so the two lines of code in bold need to be added for testing correctness of clip_quant_scale function?

The volume of the output does not become lower, which show the low volume output may cause by the output of student model instead of post processing in wavenet/parallelgen.py?

**Comment by bfs18** [comment]
@HallidayReadyOne 
1. The output distribution of the student should be close to that of the teacher. So if mu-law encoding is used in the teacher, the student should also predicts mu-law encoded signal.
3. Since "the output of the student model is scaled mu-law data", you should not apply mu-law encoding once again on that data. So the 2 lines in bold is not correct. If you would like to use the 2 lines, you should make sure that you signal is raw audio signal in range [-1, 1).

**Comment by HallidayReadyOne** [comment]
Hi @bfs18 , 

Since "the output of the student model is scaled mu-law data",  in the best case, the output of student model equal to
"
audio, _ = librosa.load('test_data/test.wav', sr=16000)
audio_mu = mu_law_numpy(audio)
audio_mu_scaled = np.asarray(audio_mu, np.float32) / (2 ** 8 / 2.)
"
So, as long as the student model is well trained, its output will be close to audio_mu_scaled?
Then execute the following code,
"
audio_int = cast_quantize_numpy(audio_mu_scaled, 2 ** 8)
audio_ = inv_mu_law_numpy(audio_int)
librosa.output.write_wav('test_data/test_inv.wav', audio_, sr=16000)
"
the volume of audio_ should be kept?

**Comment by bfs18** [comment]
Hi @HallidayReadyOne 
```
audio, _ = librosa.load('test_data/test.wav', sr=16000)
audio_mu = mu_law_numpy(audio)
audio_mu_scaled = np.asarray(audio_mu, np.float32) / (2 ** 8 / 2.)
```
is input preprocess.
```
audio_int = cast_quantize_numpy(audio_mu_scaled, 2 ** 8)
audio_ = inv_mu_law_numpy(audio_int)
librosa.output.write_wav('test_data/test_inv.wav', audio_, sr=16000)
```
is output postprocess.
As far as I am considered, the encoding decoding logic in the code is correct.The volume should be kept if everything goes well. Obviously something is wrong.

**Comment by HallidayReadyOne** [comment]
@bfs18 , 
I totally agree with you. 
My view was that the inaccurate output of student model should be the main cause of the volume drop instead of postprocess in wavenet/parallelgen.py.
I will try to solve it, progress will timely inform you.
Please let me know if you have any new idea about this problem.
Thanks.

**Comment by bfs18** [comment]
Shared deconvolution stack can reconstruct the formats between 1000 and 3000 Hz.
![share_deconv](https://user-images.githubusercontent.com/39617376/46245378-41192f80-c41f-11e8-89b7-25ee558cab5a.PNG)

When using separate deconvolution stacks, the tanh activation functions of the first 2 deconvolution stack turn to saturate, then the condition become meaningless. 
![h1](https://user-images.githubusercontent.com/39617376/46245398-876e8e80-c41f-11e8-8067-2fd6b3f9ad28.PNG)
![h2](https://user-images.githubusercontent.com/39617376/46245403-8dfd0600-c41f-11e8-80a2-4dc35c7b9b36.PNG)
![h3](https://user-images.githubusercontent.com/39617376/46245402-8d646f80-c41f-11e8-9c19-6aada5d49fb4.PNG)
![h4](https://user-images.githubusercontent.com/39617376/46245401-8d646f80-c41f-11e8-8826-791cc0443fcd.PNG)
Shared deconvolution can eliminate this problem. At least every iaf stack receives a meaningful conditional representation. 
![share_h](https://user-images.githubusercontent.com/39617376/46245400-8d646f80-c41f-11e8-808f-4c51a2656178.PNG)

**Comment by switchzts** [comment]
> Shared deconvolution stack can reconstruct the formats between 1000 and 3000 Hz.
> ![share_deconv](https://user-images.githubusercontent.com/39617376/46245378-41192f80-c41f-11e8-89b7-25ee558cab5a.PNG)
> 
> When using separate deconvolution stacks, the tanh activation functions of the first 2 deconvolution stack turn to saturate, then the condition become meaningless.
> ![h1](https://user-images.githubusercontent.com/39617376/46245398-876e8e80-c41f-11e8-8067-2fd6b3f9ad28.PNG)
> ![h2](https://user-images.githubusercontent.com/39617376/46245403-8dfd0600-c41f-11e8-80a2-4dc35c7b9b36.PNG)
> ![h3](https://user-images.githubusercontent.com/39617376/46245402-8d646f80-c41f-11e8-9c19-6aada5d49fb4.PNG)
> ![h4](https://user-images.githubusercontent.com/39617376/46245401-8d646f80-c41f-11e8-8826-791cc0443fcd.PNG)
> Shared deconvolution can eliminate this problem. At least every iaf stack receives a meaningful conditional representation.
> ![share_h](https://user-images.githubusercontent.com/39617376/46245400-8d646f80-c41f-11e8-808f-4c51a2656178.PNG)
> I also tried use the teacher's deconv stack weights to the student and not update it during training the student. The fromants are much clear than separate deconv stacks.

Why is my output a meaningless vocal? Instead of a sentence？Do you have any idea about it?

**Comment by switchzts** [comment]
@bfs18 Should I set "use_weight_norm": true to false?

**Comment by bfs18** [comment]
@switchzts  Hi, set it to False and use leaky_relu as act_func, then the net would be easier to train. I will update a stabler and easier to train default config json.

**Comment by WendongGan** [comment]
@bfs18 Thanks very much for your share! After your commmiting your latest code ( commit 3fa872b ),on Oct,22, I tried to reproduce it . I trained the teacher model. The result is better! But there is a few problems. I have trained for 608k, my batch size is 4,the loss is as follow:

image
![image](https://user-images.githubusercontent.com/40649244/47761808-f37d3480-dcf4-11e8-827f-76683843906f.png)


The result is as follow:
![image](https://user-images.githubusercontent.com/40649244/47761825-055ed780-dcf5-11e8-8da8-cd97a776d007.png)


my audio
[my-result-and-ori.zip](https://github.com/bfs18/nsynth_wavenet/files/2532430/my-result-and-ori.zip)

Where the mel spectrum is very dense, there's some noise.Have you ever had a problem like this? Do you have any idea? Will this affect the student network? Look forward for your reply and help! Thank you!  I also hope other experienced friends can discuss together。@switchzts
